Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -290,6 +290,91 @@ For white-box source code testing, you'll need a setup similar to the developer'
+
+```bash
+strings libnative-lib.so
+...
+Java_sg_vantagepoint_helloworldjni_MainActivity_stringFromJNI
+...
+Hello from C++
+...
+```
+
+##### Cross References
+
+###### Java
+Java cross references is supported by many tools, including Ghidra and [jadx](https://github.com/skylot/jadx "jadx"). With Ghidra cross references can be obtained by right clicking on the desired function in the symbol tree and selecting **Show References to**, as shown in the image below.
+
+<img src="Images/Chapters/0x05c/Ghidra_java_xrefs.png" width="500px"/>
+
+Another popular Java decompiler tool is jadx. Using jadx, cross references for a desired function can be obtained by right clicking the decompiled function and selecting **Find Usage**.
+
+<img src="Images/Chapters/0x05c/jadx_java_xrefs.png" width="500px"/>
+
+
+###### Native
+While analysing an Android application's native library using Ghidra, cross references can be obtained similarly to Java, i.e, by right clicking the desired function and selecting **Show References to**.
+
+<img src="Images/Chapters/0x05c/ghidra_native_xrefs.png" width="500px"/>
+
+
+##### API Usage
+
+Android platform provides many in-built libraries for frequently used functionalities in applications, example cryptography, Bluetooth, NFC, network or location libraries. Presence of these imported libraries in an application's code can give us valuable information about the nature of application being evaluated.
+
+For instance, if in an application `javax.crypto.Cipher` class is imported, it indicates that application will be performing some kind of cryptographic operation. Fortunately, Cryptographic calls are very standard in nature, i.e, they need to be called in a particular order to work correctly, this knowledge can be helpful when analysing cryptography APIs. Example, by looking for `Cipher.getInstance()` function, we can determine the cryptographic algorithm being used. With such approach we can directly move to analysing cryptographic assets, which often are very critical in an application. Further information on how to analyse Android's cryptographic APIs is discussed in  "[Android Cryptographic APIs](0x05e-testing-cryptography "Android Cryptographic APIs")" section.
+
+Similarly, the above approach can be used in determining where and how an application is using NFC. For instance, an application using Host-based Card Emulation for performing digital payments, must use `android.nfc` package. By searching for critical functions like `processCommandApdu()` from `android.nfc.cardemulation.HostApduService` class, can provide us a starting point for NFC API analysis.
+
+Checking for the imported classes in an application can help us in understanding the functionalities implemented. Further by determining the most important functions of an imported class, one can start analysis by assuming this function as the centrepiece of the puzzle.
+
+
+##### Checking Secure Connections
```

Comment:
I think this section can be removed. 
Or if we keep it we should make a note that you are getting much more reliable results when doing dynamic testing with Burp. Not all communication is via HTTP, but also plain tcp packets or XMPP. This can easily be observered in dynamic testing, which can be very tedious to identify in RE. 

@cpholguera What do you think?

Reply:
Those are very good points @sushi2k , I'd keep the section as it's purpose is exactly that: to show what you can statically test regarding secure connections. If there's not so much you can test or there are better ways do do it, this is _the place_ to describe that.

I'd suggest @su-vikas can consider your comments above and enhance the section.

Also I'd think of a little checklist of what can/should be tested statically: e.g. trying to find out which protocols are being used, checking the network configuration files, pinned certs vs target domains, more ideas? so that one can ensure that at least everything that can be checked was checked and often this minimal info also helps ease (servers as input for) the dynamic analysis ;)

Reply:
Agree!